Items

Mryle opened the pull request:
Btw, nie mam pojęcia czemu automatic merge nie przechodzi.

Mryle committed:
Adding Item UI p1

Also started adding classes for items, and minor changes in scripts.

Mryle committed:
Fully usable Item UI

As title + example item (not working). Items aren't dropped properly (just
deleted).

Mryle committed:
Refactored: Item - component

Also some minor twaks.
Dropping items should be working.

Mryle committed:
Merge branch 'master' into items

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+            {
+                equipped[1] = obj;
+                equipped[1].onPickUp(player.gameObject, this);
+                equipped[1].onEquip(0);
```
Shouldn't it be `equipped[1].onEquip(1)`?

Mryle replied:
Yes, it should be. Typo.

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+
+	// Use this for initialization
+	void Start () {
+        images = new UnityEngine.UI.Image[8];
```
Why is it done on Start() and not when declaring member? Also maybe we could extract the constant out.

Xanewok commented on Assets/Scripts/Items/ItemObject.cs:
```diff
@@ -0,0 +1,75 @@
+using UnityEngine;
+
+/*
+    Nadklasa dla przedmiotów.
```
I'm not sure I get it 😢
Also probably should be in english to stay consistent (I guess?)

Xanewok commented on Assets/Scripts/Items/ItemObject.cs:
```diff
@@ -0,0 +1,75 @@
+
+    public virtual bool canBePicked(GameObject player)
+    {
+        return pickUpDelay == 0f;
```
In this case we manually clamp this to 0 so it's ok, but in general for floating point numbers it's generally safer to check for `<= Mathf.Epsilon` or at least `<= 0f`

Xanewok commented on Assets/Scripts/Items/ItemObject.cs:
```diff
@@ -0,0 +1,75 @@
+        return pickUpDelay == 0f;
+    }
+
+    public virtual Sprite getSprite()
```
Shouldn't this be `abstract`?

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+    void ShowInventory()
+    {
+        canvas.scaleFactor = Mathf.Lerp(canvas.scaleFactor, 1.0f, Time.deltaTime * fadeInScale);
+        if (canvas.scaleFactor > 0.2f)
```
Why check for `scaleFactor` and why `0.2` in particular?

Xanewok commented on Assets/Scripts/FollowingCamera.cs:
```diff
@@ -0,0 +1,46 @@
+﻿using System.Collections;
```
We do have a script already for a following camera. Not sure if having an externally controlled camera is good? We could just make appropriate camera render to different viewports if we want to have different following cameras for different players.

Mryle replied:
I suspect that this script was a result of a faulty rebase. Will investigate.

Xanewok commented on Assets/Scripts/FollowingCamera.cs:
```diff
@@ -0,0 +1,46 @@
+    public void SetFollowingCamera(GameObject obj)
+    {
+        
+        followingCamera = obj == null ? null : obj;
```
`followingCamera = obj` would be enough

Xanewok commented on Assets/Scripts/FollowingCamera.cs:
```diff
@@ -0,0 +1,46 @@
+    {
+        if (followingCamera == null)
+        {
+            GameObject camera = GameObject.FindGameObjectWithTag("MainCamera");
```
Not sure if this is for many cameras/players anymore if we get only `MainCamera`

Xanewok commented on Assets/Scripts/Items/ItemObject.cs:
```diff
@@ -0,0 +1,75 @@
+    {
+        if (pickUpDelay > 0f)
+        {
+            pickUpDelay -= Time.deltaTime;
```
This could be replaced with `pickUpDelay = Mathf.Max(0f, pickUpDelay - Time.deltaTime)` if that's all it should do

Mryle replied:
Well yes, haven't thought of it.

Xanewok commented on Assets/Scripts/PlayerControlls.cs:
```diff
@@ -85,4 +86,14 @@ void Update()
         }
     }
+
+    public Vector3 getLastTurn()
```
Can't we just get a transform.rotation instead? Or will it be used explicitly when interfacing externally with `PlayerControlls`?

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+        canvas.scaleFactor = Mathf.Lerp(canvas.scaleFactor, 1.0f, Time.deltaTime * fadeInScale);
+        if (canvas.scaleFactor > 0.2f)
+            canvas.enabled = true;
+        Vector3 aim = player.getLastTurn();
```
Maybe we should disconnect actual player movement for when inventory menu is brought up and player is managing items? I suspect it might be ideal for player to change his items mid-fight for example or if he'd be using item that needs player orientation (such as flashlight).

Mryle replied:
I suppose yes, but for that we a better class for managing it. As for right now, me might leave it as is.

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+        Vector3 aim = player.getLastTurn();
+        Vector2 aim2d = Vector2.zero;
+        aim2d.Set(aim.x, aim.z);
+        if (aim2d.SqrMagnitude() > 0.5)
```
When is it going to be > 0.5?

Mryle replied:
In general it was only used as a kind of failsafe mechanism. It is to be redeclared, so probably wont be needed.

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+            {
+                angle = -angle;
+            }
+            int kol = (int)(angle + 22.5f);
```
I'm not sure I follow the explicit logic here :cry:

Mryle replied:
Just a way to count angles bigger than 180 degree. As far as i know, there is no explicit function that would do such a thing

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+            canvas = gameObject.transform.GetComponentInChildren<Canvas>();
+        }
+        equipped = new ItemObject[2];
+        picked = new ItemObject[7];
```
Maybe we can extract the constants?

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+    {
+        if (canvas.scaleFactor > 0.2)
+        {
+            canvas.scaleFactor = Mathf.Lerp(canvas.scaleFactor, 0.1f, Time.deltaTime * fadeOutScale);
```
Having scale range of [0.1f;0.2f] should probably be exposed somewhere (either externally or as a constant)

Xanewok commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,193 @@
+using UnityEngine;
+
+/**
+    Klasa ekwipunku dla gracza.
```
Like above, I think that comments should stay in english to be consistent

Mryle replied:
Ok, will be changed.

Xanewok reviewed (changes requested):
Looks good overall except some details and duplicated camera follow script.

Mryle committed:
Changes specified in PR1

Changed commentary to english.
Rewrite how slots are selected (using players rotation)
Rewrite items superclass (it contains useful things)
Corrected some typos.

Mryle committed:
Last changes for items.

Updated Inventory prefab and testing scene.

Xanewok commented:
Looks good 👍
